distogram/__init__.py

```diff
@@ -1,240 +1,203 @@
-        diff2 = _weighted_diff(h, bins[index], (new_value, 1))
-        if diff1 < diff2:
-            diff = diff1
-            i_bin = index-1
-        else:
-            diff = diff2
-            i_bin = index
+        diff1 = _weighted_diff(h, (new_value, 1), h.bins[index - 1])
+        diff2 = _weighted_diff(h, h.bins[index], (new_value, 1))
+
+        i_bin, diff = (index - 1, diff1) if diff1 < diff2 else (index, diff2)
+
+        return i_bin if diff < h.min_diff else -1
 
-        if diff < min_diff:
-            return i_bin
-        else:
-            return -1
     return -1
 
 
-def update(h, value, count=1):
-    '''Adds a new element to the distribution.
+def update(h: Distogram, value: float, c: int = 1) -> Distogram:
+    """ Adds a new element to the distribution.
 
     Args:
         h: A Distogram object.
         value: The value to add on the histogram.
-        count: [Optional] The number of times that value must be added.
+        c: [Optional] The number of times that value must be added.
 
     Returns:
         A Distogram object where value as been processed.
-    '''
-    bins = h.bins
+    """
     index = 0
-    if len(bins) > 0:
-        if value <= bins[0][0]:
+    if len(h.bins) > 0:
+        v0, _ = h.bins[0]
+        vl, _ = h.bins[-1]
+
+        if value <= v0:
             index = 0
-        elif value >= bins[-1][0]:
+        elif value >= vl:
             index = -1
         else:
-            if len(bins) >= h.bin_count:
-                index = _linear_index(bins, value)
-            if index == 0:
-                index = bisect_left(bins, (value, 1))
+            index = bisect_left(h.bins, (value, 1))
 
-    if index > 0 and len(bins) >= h.bin_count:
-        in_place_index = _search_in_place_index(h, value, index)
-        if in_place_index > 0:
-            h = _trim_in_place(h, value, count, in_place_index)
+        vi, fi = h.bins[index]
+        if vi == value:
+            h.bins[index] = (vi, fi + c)
             return h
 
-    if len(bins) > 0:
-        bin = bins[index]
-        if bin[0] == value:
-            bin = (bin[0], bin[1]+count)
-            h.bins[index] = bin
+    if index > 0 and len(h.bins) >= h.bin_count:
+        in_place_index = _search_in_place_index(h, value, index)
+        if in_place_index > 0:
+            h = _trim_in_place(h, value, c, in_place_index)
             return h
 
     if index == -1:
-        bins.append((value, count))
+        h.bins.append((value, c))
         if h.diffs is not None:
             diff = _weighted_diff(h, h.bins[-1], h.bins[-2])
             h.diffs.append(diff)
-            if diff < h.min_diff:
-                h.min_diff = diff
+            h.min_diff = min(h.min_diff, diff)
     else:
-        bins.insert(index, (value, count))
+        h.bins.insert(index, (value, c))
         if h.diffs is not None:
-            h.diffs.insert(index, None)
+            h.diffs.insert(index, 0)
```

Comment:
Having the `None` here threw me for a while. I think that diffs should always be a list of floats because `_update_diffs` never checks whether a diff is None. If we want to leave `None` here, then we should check for that in `_update_diffs` otherwise we could run into bugs.

Reply:
I'm fairly certain that the update always happens. If we can find a failing test case where the update does not happen, we should fix it